### PR TITLE
🚀 Create Qodana TypeScript GitHub Action

### PR DESCRIPTION
**Pull request by tiulpin**
Added the initial TypeScript version of our GitHub Action.
It can
- run Qodana images (as before)
- automatically upload the report (so the logs won't be lost anymore)
- automatically set up GitHub caches
- add annotations

Every added step is enabled by default but can be easily disabled by action parameters. Note: I changed `action.yaml` – let's discuss it (I believe we should be able to allow any Docker image names)

- Tests (here I'm testing things I can locally and via provided GitHub Action to run the Action in Actions 🤯) – https://github.com/tiulpin/qodana-action/runs/4511564215
- Example run – https://github.com/tiulpin/dokker/runs/4511638796?check_suite_focus=true
- Annotations preview in Pull Request view – https://github.com/tiulpin/dokker/pull/2/files

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,99 @@
+# Dependency directory
```
Comment (ivan-infro) [inline .gitignore:1]: It seems to me, there are "dist" folder missing

File: src/annotations.ts [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+ * @param path The path to the SARIF file to publish.
+ */
+export async function publishAnnotations(
+  token: string,
+  path: string
+): Promise<void> {
+  try {
+    const output = parseSarif(path)
+    if (output.annotations.length >= MAX_ANNOTATIONS) {
+      for (let i = 0; i < output.annotations.length; i += MAX_ANNOTATIONS) {
```
Comment (ivan-infro) [inline src/annotations.ts:228]: Only first iteration will be executed

**Review by ivan-infro — APPROVED** [review]

File: README.md [hunk 3]
```diff
@@ -8,45 +8,54 @@ Qodana already supports PHP, Java, and Kotlin projects, and will eventually supp
...
+    fail-threshold: 10  # to set a quality gate to the number of found problems
     additional-env-variables: |
         IDEA_PROPERTIES='/data/project/idea.properties'
 ```
 
 Before you begin, view the list of [Supported Technologies](https://www.jetbrains.com/help/qodana/supported-technologies.html). For the full documentation of the action's inputs, see [action.yaml](action.yaml).
 
+## Configuration
+
+- `linter`: Qodana linter. All possible values can be found here: https://hub.docker.com/r/jetbrains/qodana. Is required. Defaults to `jetbrains/qodana-jvm-community`
```
Comment (a-kuznetsova) [inline README.md:33]: Let's change `Is required.` to `Required.`

File: README.md [hunk 4]
```diff
@@ -8,45 +8,54 @@ Qodana already supports PHP, Java, and Kotlin projects, and will eventually supp
...
+- `upload-result`: Upload Qodana results as an artifact to the job. Optional. Defaults to `true`
+- `use-caches`: Utilize GitHub caches for Qodana runs. Optional. Defaults to `true`
+- `use-annotations`: Use annotation to mark the results in GitHub user interface. Optional. Defaults to `true`
+- `github-token`: GitHub token to be used for uploading results. Optional. Defaults to `${{ github.token }}`
+
 ## License
 
-View [license information](https://www.jetbrains.com/legal/?fromFooter#licensing) for the Qodana for JVM Community linter software contained in this image.
+This repository contains source code for Qodana GitHub Action and is licensed under [Apache-2.0](./LICENSE).
+View [license information](https://www.jetbrains.com/legal/?fromFooter#licensing) for the Qodana images.
```
Comment (a-kuznetsova) [inline README.md:57]: Qodanas' EAP images are released under different license. 

```
Using Qodana docker image you agree to [JetBrains EAP user agreement](https://www.jetbrains.com/legal/docs/toolbox/user_eap/) and [JetBrains privacy policy](https://www.jetbrains.com/legal/docs/privacy/privacy/). The docker image includes an evaluation license which will expire in 30-day. Please ensure you pull a new image on time. 
``` from https://github.com/JetBrains/qodana-action-eap

File: README.md [hunk 5]
```diff
@@ -8,45 +8,54 @@ Qodana already supports PHP, Java, and Kotlin projects, and will eventually supp
...
-* `additional-volumes` - Additional volumes to mount to qodana docker image
-* `additional-env-variables` - Additional environment variables to pass to qodana docker image
-
 ```yaml
 - uses: JetBrains/qodana-action@v3.2.1
-  with:
-    linter: qodana-jvm-community
-    fail-threshold: 10
+  with:  # optional
+    linter: jetbrains/qodana-jvm-community:2021.2  # Docker image full name
```
Comment (a-kuznetsova) [inline README.md:23]: I think we should change the tag to `latest` or `2021.3-eap`

**Comment by a-kuznetsova** [comment]
LGTM

**Review by a-kuznetsova — APPROVED** [review]

**Pull request merged**